lattice-controller/src/client_cache.rs

```diff
@@ -0,0 +1,248 @@
+    let auction_timeout = Duration::from_millis(config.auction_timeout_ms);
+    let lattice_prefix = config.lattice_prefix.clone();
+    let conn = connect(config).await?;
+    let client = wasmcloud_control_interface::Client::new(
+        conn,
+        Some(lattice_prefix),
+        timeout,
+        auction_timeout,
+    );
+    Ok(client)
+}
+
+/// Create a new nats connection
+async fn connect(cfg: &ConnectionConfig) -> RpcResult<async_nats::Client> {
+    let cfg = cfg.clone();
+    let opts = match (cfg.auth_jwt, cfg.auth_seed) {
+        (Some(jwt), Some(seed)) => {
+            let key_pair = std::sync::Arc::new(
+                KeyPair::from_seed(&seed)
+                    .map_err(|e| RpcError::ProviderInit(format!("key init: {}", e)))?,
+            );
+            async_nats::ConnectOptions::with_jwt(jwt, move |nonce| {
+                let key_pair = key_pair.clone();
+                async move { key_pair.sign(&nonce).map_err(async_nats::AuthError::new) }
+            })
+        }
+        (None, None) => async_nats::ConnectOptions::default(),
+        _ => {
+            return Err(RpcError::InvalidParameter(
+                "must provide both jwt and seed for jwt authentication".into(),
+            ));
+        }
+    };
+    if cfg.cluster_uris.is_empty() {
+        return Err(RpcError::NotInitialized(
+            "No NATS URIs supplied".to_string(),
+        ));
+    }
+
+    let url = cfg.cluster_uris.get(0).unwrap();
+
+    let conn = opts
+        .event_callback(|event| async move {
+            // lattice prefix/ID will already be on the span from earlier calls
+            match event {
+                async_nats::Event::Disconnect => debug!("NATS client disconnected"),
+                async_nats::Event::Reconnect => debug!("NATS client reconnected"),
+                async_nats::Event::ClientError(err) => debug!("NATS client error occured: {}", err),
+                other => debug!("NATS client other event occurred: {}", other),
+            }
+        })
+        .connect(url)
+        .await
+        .map_err(|e| RpcError::ProviderInit(format!("Nats connection to {}: {}", url, e)))?;
+
+    Ok(conn)
+}
+
+/// Discovers a list of expired (access time within grace period) connections
+/// and then removes them from the cache.
+async fn evacuate_cache(
+    m: Arc<RwLock<HashMap<String, ClientMetadata>>>,
+    c: Arc<RwLock<HashMap<String, Client>>>,
+    period: Duration,
+) {
+    let expired_keys: Vec<String> = {
+        let meta = m.read().await;
+        let now = Instant::now();
+
+        meta.iter()
+            .filter(|(_k, v)| now.duration_since(v.last_accessed) > period)
+            .map(|(k, _v)| k.to_string())
+            .collect()
+    };
+
+    if !expired_keys.is_empty() {
+        trace!(
+            "Removing NATS clients from cache: {}",
+            expired_keys.join(",")
+        );
+    }
+
+    let mut conns = c.write().await;
+    conns.retain(|k, _v| !expired_keys.contains(k));
```

Comment:
nit: this does a full table scan and is O(n) for n total items. If len(expired_keys) << len(conns) then it would be more efficient to delete them individually

Reply:
it does a full table scan in read mode and then lets go of the lock. this is a deliberate choice to not hold a write lock for any longer than necessary.

Reply:
I'm going to pass on optimizing this for now because I am pretty sure we're going to ditch this cache in favor of another crate, or at the very least do some other refactoring.